### ISSUE TITLE
GAT in Allocator Trait

### ISSUE THREAD

**Issue by benjamin-lieser** [dimforge/nalgebra#1396]
Now that GAT are stable in Rust, they could improve the Allocator Trait.

The `T` should be a GAT, which would remove a generic argument, and would allow the removal of a lot of trait bounds.

Now we have a lot of code like this:
```
impl<T: ComplexField, D> OMatrix<T, D, D>
where
    D: DimMin<D, Output = D>,
    DefaultAllocator: Allocator<T, D, D>
        + Allocator<(usize, usize), DimMinimum<D, D>>
        + Allocator<T, D>
        + Allocator<T::RealField, D>
        + Allocator<T::RealField, D, D>,
```

`T` can then be late bound when actually allocating.

I am not sure if I have overlooked something. There might be bad interactions for `T : ?Sized`

**Comment by sebcrozet** [comment]
Thank you for the heads up! I made a PR with that change.

I didn’t realize that GAT were stabilized for arbitrary type parameters. I thought it was only for lifetimes `type AssociatedType<'a>`. Do you know what version of the compiler stabilized that for actual type parameters?

**Comment by benjamin-lieser** [comment]
I am pretty sure they got stabilized together with the lifetimes in 1.65
It's definitely working with the current compiler: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=084a9df304548bbc6021da2fcce1c31a

**Comment by Andlon** [comment]
I believe this was resolved with #1397 being merged.
